### PR TITLE
Pagination in quick view

### LINKED ISSUE #455: Add pagination in quick view
This issue intends to keep track of the changes in PR #405 about implementing pagination in the quick view.

Belongs to the set of priorities in #330.

### PR DESCRIPTION
**Pull request by LuiggiTenorioK**
Closes #455 

**Part of #330** 

Uses API changes from https://github.com/BSC-ES/autosubmit-api/pull/220

The idea of this PR is to allow a partial load of the jobs in the quick view. This will implement:

- [x] Handling v4 endpoint that allows pagination, and querying instead of loading the whole job list
- [x] Styling job items of the list in the client and not depending on the backend
- [x] Keep search params persistent in the URL to facilitate sharing and persist state

Some drawbacks of this implementation include:

- Dropping the wrappers data in the job list
- Dropping the job counters in the status selection
- Performance is only optimized if the experiment uses the DB version of the job list and not the PKL one

### CODE REVIEW

**Comment by pablogoitia** [comment]
Hi, @LuiggiTenorioK! I see that the implementation is well underway, perhaps just waiting for the job endpoint to be finished, along with some final adaptations, tweaks and testing. Here are a few questions I have:

-----------------

> Dropping the wrappers data in the job list

I understand that this is a feature we didn't have until now and that, due to its complexity, we've decided to postpone it. Is this a “nice-to-have” or something we should address?

> Dropping the job counters in the status selection

I see that when I apply the status filter to the query, it returns the number of jobs it found with that status and updates the counter. I think it's a great solution and more than enough.

-----------

On the other hand, do you think I should add another kind of elements such as a drop-down list to increase the number of jobs displayed per page?

Can you remember if there was any specific case, element... you were concern about in this new implementation of the view? I have no notes for this issue, so... that's simply nice. I will test it anyway just in case

**Comment by LuiggiTenorioK** [comment]
> I understand that this is a feature we didn't have until now and that, due to its complexity, we've decided to postpone it. Is this a “nice-to-have” or something we should address?

It is a "nice-to-have" for now. Eventually, we need to find a way to show the wrappers quickly without adding significant time complexity to this view.



> I see that when I apply the status filter to the query, it returns the number of jobs it found with that status and updates the counter. I think it's a great solution and more than enough.

It is good, but ensure that it is clear to the user that only some of the counters are calculated and the others are not, and how to calculate them.



> On the other hand, do you think I should add another kind of elements such as a drop-down list to increase the number of jobs displayed per page?

Is a good proposal and not hard to implement :+1: 




> Can you remember if there was any specific case, element... you were concern about in this new implementation of the view? I have no notes for this issue, so... that's simply nice. I will test it anyway just in case

I was only concerned about the wrappers visualization. In both the quick view and tree view.

File: src/pages/ExperimentQuick.jsx [hunk 1]
```diff
@@ -150,14 +104,34 @@ const ExperimentQuick = () => {
...
+
+    return size
+  }, [searchParams])
+
+  const { data, isFetching, refetch } = autosubmitApiV4.endpoints.getExperimentJobs.useQuery({
+    expid: routeParams.expid,
+    page: currentPage,
+    page_size: pageSize,
+    status: searchParams.get("status") || undefined,
+    query: searchParams.get("query") || undefined
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx]: Hi @pablogoitia, I'm resuming the implementation from the API side; nearly finished.
I was testing the integration with the gui and I think the `query` param should be `job_name` instead. Currently FastAPI is ignoring `query`.
I'll implement the response model to make it more explicit from the API side.

Comment (pablogoitia) [reply src/pages/ExperimentQuick.jsx]: Hi @ntorqulu. Thanks for pointing it out. You'll see it corrected in the upcoming commits :)
I'll deeply review the PR soon in order to check if there is something else pending in Luiggi's implementation and if my changes work smoothly

File: src/pages/ExperimentQuick.jsx [hunk 2]
```diff
@@ -150,14 +104,34 @@ const ExperimentQuick = () => {
...
+  }, {
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx:196]: The API exposes an optional `view`. I would send `view: quick". The quick view only renders `name` + `status`, so the requests can save in the payload. 
```suggestion
 job_name: searchParams.get("job_name") || undefined,,
 view: "quick"
```

File: src/pages/ExperimentQuick.jsx [hunk 3]
```diff
@@ -172,31 +146,42 @@ const ExperimentQuick = () => {
...
+      ...Object.fromEntries(searchParams.entries()),
+      page: 1,
+      page_size: newSize
+    })
+  }
+
+  const [queryInput, setQueryInput] = useState(searchParams.get("query") || "")
+
+  const handleFilterClick = () => {
+    const { query, ...rest } = Object.fromEntries(searchParams.entries())
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx]:
```suggestion
    const { job_name, ...rest } = Object.fromEntries(searchParams.entries())
```

File: src/pages/ExperimentQuick.jsx [hunk 4]
```diff
@@ -172,31 +146,42 @@ const ExperimentQuick = () => {
...
+    setSearchParams({
+      ...rest,
+      page: 1,
+      ...(queryInput && { query: queryInput })
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx]:
```suggestion
      ...(jobNameInput && { job_name: jobNameInput })
```

File: src/pages/ExperimentQuick.jsx [hunk 5]
```diff
@@ -172,31 +146,42 @@ const ExperimentQuick = () => {
...
+  const handlePageSizeChange = (e) => {
+    const newSize = parseInt(e.target.value, 10)
+    setSearchParams({
+      ...Object.fromEntries(searchParams.entries()),
+      page: 1,
+      page_size: newSize
+    })
+  }
+
+  const [queryInput, setQueryInput] = useState(searchParams.get("query") || "")
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx]:
```suggestion
  const [jobNameInput, setJobNameInput] = useState(searchParams.get("job_name") || "")
```

File: src/pages/ExperimentQuick.jsx [hunk 6]
```diff
@@ -209,48 +194,47 @@ const ExperimentQuick = () => {
...
+        <div className="grow flex">
           <input
-            value={filters.filter}
-            onChange={handleFilterChange}
-            className="form-input w-full"
+            className="form-input w-full rounded-r-none"
             placeholder="Filter job..."
-          />
-          {/* <button className="btn btn-dark fw-bold px-4">Filter</button> */}
+            value={queryInput}
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx]:
```suggestion
            value={jobNameInput}
```

File: src/pages/ExperimentQuick.jsx [hunk 7]
```diff
@@ -209,48 +194,47 @@ const ExperimentQuick = () => {
...
+            onChange={(e) => setQueryInput(e.target.value)}
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx]:
```suggestion
            onChange={(e) => setJobNameInput(e.target.value)}
```

File: src/services/autosubmitApiV4.js [hunk 8]
```diff
@@ -46,10 +46,16 @@ export const autosubmitApiV4 = createApi({
...
+            query: ({ expid, query = undefined, status = undefined, page = undefined, page_size = undefined }) => {
                 return {
                     url: `experiments/${expid}/jobs`,
-                    method: "GET"
+                    method: "GET",
+                    params: {
+                        query,
+                        status,
+                        page,
+                        page_size
```
Comment (ntorqulu) [inline src/services/autosubmitApiV4.js]:
```suggestion
                        job_name,
                        status,
                        page,
                        page_size,
                        view
```

**Review by ntorqulu — COMMENTED** [review]
Added just a few comments related to the `job_name` and a faster rendering selecting the view type. Other than that, no other issues found!
The pr from the API side is ready too, tested it with the latest commit.

**Comment by pablogoitia** [comment]
Hi, @ntorqulu. Thanks for the update! Then I will start reviewing this :) I forgot to tell you that yesterday tried changing query for jobName and worked smoothly, so thanks again for pointing that out! I will use your suggested changes to review my local ones (sorry because if I had pushed a commit with the changes maybe you might have saved some time)

**Comment by pablogoitia** [comment]
Here I propose two alternatives for selecting the page size (in jobs):

### First option: integrated within the job counter

image


### Second option: below the page selector

image

Using one or another is inexpensive for me. The second was the original I prepared for this PR some weeks ago, and the first is a quick draft because I found the second to occupy unnecessary space.

**Comment by ntorqulu** [comment]
Hi @pablogoitia! Both look really nice, but I would go with the first one (integrated in the job counter).
I can imagine users seeing the number of jobs displayed per page and wanting to change it, but not finding where to do it because it's in the bottom of the page.

**Comment by pablogoitia** [comment]
>  I can imagine users seeing the number of jobs displayed per page and wanting to change it, but not finding where to do it because it's in the bottom of the page.

Nice! Glad you also like it the most. Also, when the count of jobs is lower than the default page size, the selector is replaced by the count.

**Comment by pablogoitia** [comment]
Hi, @ntorqulu. The PR is now ready for review.

Note that I have changed lots of things and now the component differs a lot with respect to the base, and some things may have varied since your last use of the branch.

Maybe you'll be surprised because of the tests. I developed them because this was the only view (as far as I know) that had not a single test, and I felt it was necessary to have a few (they are not, actually). I think they are so complete, and helped a lot to find edge cases that were failing.

Feel free to take it a look whenever you can

File: src/pages/ExperimentQuick.jsx [hunk 9]
```diff
@@ -1,15 +1,25 @@
...
+
+const ITEMS_PER_PAGE_OPTIONS = [DEFAULT_ITEMS_QUICK_VIEW, 500, 1000];
 
 const QuickJobList = ({ jobs, onSelectionChange }) => {
+  const [selectedJobIds, setSelectedJobIds] = useState(new Set());
+  const [lastClickedIndex, setLastClickedIndex] = useState(null);
+
+  useEffect(() => {
+    onSelectionChange(selectedJobIds);
+  }, [selectedJobIds, onSelectionChange]);
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx:21]: One question about the jobs selection in multiple pages: which should be the behaviour?
Currently if i select job A on page 1 and go to page 2 the job is deselected. Is it intended?

Also, found some strange behaviors with job selection across multiple pages.
I attach a video, not sure if it's worth fixing:

https://github.com/user-attachments/assets/ba0f0646-c4b9-4fb2-b391-8dad92733462

If I select multiple jobs from one page, then select others from another page and go back to the first page it selects all the jobs from the multiple pages...

File: src/pages/ExperimentQuick.jsx [hunk 10]
```diff
@@ -168,92 +220,104 @@ const ExperimentQuick = () => {
...
-          Showing {jobs.length} of{" "}
-          <strong>
-            {data && data.tree_view && data.tree_view.length} total jobs
-          </strong>
+        <div className="flex items-center gap-1 text-sm" style={{ whiteSpace: "nowrap" }}>
+          <span>Showing</span>
+          {data?.pagination?.total_pages == 1 ? (
+            <strong>{data?.pagination?.page_items || "0"}</strong>
+          ) : (
+            <select id="jobs-per-page" value={pageSize} onChange={handlePageSizeChange}
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx]: On the page-size selector, once a filter allows all jobs to fit into one page the user cannot change back the number of jobs per page without editing the url.
Could be a feature or a bug, not sure xd.

https://github.com/user-attachments/assets/32a01baf-7bb6-4430-9001-6557dd2ab033

File: src/pages/ExperimentQuick.jsx [hunk 11]
```diff
@@ -168,92 +220,104 @@ const ExperimentQuick = () => {
...
-              READY
-            </option>
-            <option value="WAITING">
-              WAITING
-            </option>
-            <option value="SUBMITTED">
-              SUBMITTED
-            </option>
+            <option value="" className="bg-white text-black">Any status</option>
+            {JOB_STATUSES.map((status) => (
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx:281]: Not for this pr, but as a follow-up issue I would extract a shared component and use it in `ExperimentQuick.jsx` and `ExperimentGraph.jsx`.
This way the visualization is not browser-dependent.

https://react-aria.adobe.com/ListBox  can guarantee the colored statuses on all browsers, could be a good alternative.

File: cypress/e2e/quick_view.cy.js [hunk 12]
```diff
@@ -0,0 +1,649 @@
...
+  it("check page size selector when there is just a single page", () => {
+    cy.intercept(
+      {
+        method: "GET",
+        pathname: `/v4/experiments/${expid}/jobs`,
+      },
+      {
+        statusCode: 200,
+        body: {
+          jobs: [],
```
Comment (ntorqulu) [inline cypress/e2e/quick_view.cy.js:311]: With one job?

File: cypress/support/utils.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
+export const JOB_STATUSES = [
```
Comment (ntorqulu) [inline cypress/support/utils.js:1]: Can be re-exported from `../../src/services/utils` to avoid having duplicated code and drifts from the implementation.
```
export {
    JOB_STATUSES,
    DEFAULT_ITEMS_QUICK_VIEW,
    DEFAULT_STATUS_QUICK_VIEW,
} from "../../src/services/utils";
```

File: src/services/utils.js [hunk 14]
```diff
@@ -42,6 +42,21 @@ export const saveSVGObj = (svgData, filename) => {
...
+  "HELD",
+  "QUEUING",
+  "RUNNING",
+  "COMPLETED",
+  "FAILED",
+  "SUSPENDED",
+  "SKIPPED",
+  "DELAYED"
+]
+
```
Comment (ntorqulu) [inline src/services/utils.js:60]: I would define here the constants `DEFAULT_ITEMS_QUICK_VIEW`, `ITEMS_PER_PAGE_OPTIONS` and `DEFAULT_STATUS_QUICK_VIEW` so that cypress specs can import them.

File: src/consts.js [hunk 15]
```diff
@@ -1,7 +1,7 @@
 export const AUTOSUBMIT_API_SOURCE = import.meta.env.REACT_APP_AUTOSUBMIT_API_SOURCE;
 export const PUBLIC_URL = import.meta.env.BASE_URL
 
-export const MAX_ITEMS_QUICK_VIEW = 100;
+export const DEFAULT_ITEMS_QUICK_VIEW = 100;
```
Comment (ntorqulu) [inline src/consts.js:4]: And delete it from here.

File: src/services/autosubmitApiV4.js [hunk 16]
```diff
@@ -46,10 +46,17 @@ export const autosubmitApiV4 = createApi({
             }
         }),
         getExperimentJobs: builder.query({
-            query: ({ expid }) => {
+            query: ({ expid, view = undefined, job_name = undefined, status = undefined, page = undefined, page_size = undefined }) => {
```
Comment (ntorqulu) [inline src/services/autosubmitApiV4.js]: RTK Query already drops undefined parameters, no need to default it to undefined.

File: src/services/autosubmitApiV4.js [hunk 17]
```diff
@@ -201,9 +208,9 @@ export const autosubmitApiV4 = createApi({
...
                 url: `/runners/command/get-runner-run-status`,
                 method: "POST",
-                body: { 
+                body: {
                     expid: expid
-                 }
+                }
             })
         }),
         runnerCreateNewExperiment: builder.mutation({
```
Comment (ntorqulu) [inline src/services/autosubmitApiV4.js:223]: From `autosubmitApiV3`, `useGetExperimentQuickViewQuery` is no longer used. Could be removed.

File: src/pages/ExperimentQuick.jsx [hunk 18]
```diff
@@ -1,15 +1,25 @@
...
 import useASTitle from "../hooks/useASTitle";
 import useBreadcrumb from "../hooks/useBreadcrumb";
-import { cn } from "../services/utils";
+import { cn, getStatusBadgeStyle, JOB_STATUSES } from "../services/utils";
 import { ChangeStatusModal } from "../common/ChangeStatusModal";
 import BottomPanel from "../common/BottomPanel";
 import FetchJobDetailCard from "../common/FetchJobDetailCard";
+import Paginator from "../common/Paginator";
+
+const ITEMS_PER_PAGE_OPTIONS = [DEFAULT_ITEMS_QUICK_VIEW, 500, 1000];
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx:13]: Delete the constant definition, import it from services/utils

File: src/pages/ExperimentQuick.jsx [hunk 19]
```diff
@@ -1,15 +1,25 @@
-import { useParams } from "react-router-dom";
-import { useGetExperimentQuickViewQuery } from "../services/autosubmitApiV3";
-import { useEffect, useState } from "react";
-import { MAX_ITEMS_QUICK_VIEW } from "../consts";
+import { useParams, useSearchParams } from "react-router-dom";
+import { autosubmitApiV4 } from "../services/autosubmitApiV4";
+import { useEffect, useState, useMemo } from "react";
+import { DEFAULT_ITEMS_QUICK_VIEW } from '../consts';
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx:4]: Import it from services/utils

File: src/pages/ExperimentQuick.jsx [hunk 20]
```diff
@@ -168,92 +220,104 @@ const ExperimentQuick = () => {
...
-            <option value="READY">
-              READY
-            </option>
-            <option value="WAITING">
-              WAITING
-            </option>
-            <option value="SUBMITTED">
-              SUBMITTED
-            </option>
+            <option value="" className="bg-white text-black">Any status</option>
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx]: Change `Any status` for `DEFAULT_STATUS_QUICK_VIEW`

File: src/pages/ExperimentQuick.jsx [hunk 21]
```diff
@@ -168,92 +220,104 @@ const ExperimentQuick = () => {
...
+        <div className="flex items-center gap-1 text-sm" style={{ whiteSpace: "nowrap" }}>
+          <span>Showing</span>
+          {data?.pagination?.total_pages == 1 ? (
+            <strong>{data?.pagination?.page_items || "0"}</strong>
+          ) : (
+            <select id="jobs-per-page" value={pageSize} onChange={handlePageSizeChange}
+              className="form-select border border-primary text-primary dark:bg-primary dark:text-white font-bold">
+              {ITEMS_PER_PAGE_OPTIONS.map((option) => (
+                <option key={option} value={option}>
+                  {option === DEFAULT_ITEMS_QUICK_VIEW ? `${option}` : option}
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx]: Dead ternary?
```suggestion
                  {option}
```

**Review by ntorqulu — COMMENTED** [review]
Hi @pablogoitia, review done!
Overall, great job, the pagination really improves the loading time.
I think the job selection is a little bit bugged, so we should agree which is the expected behavior of selecting jobs across different pages.

File: src/pages/ExperimentQuick.jsx (hunk 9 above)
Comment (pablogoitia) [reply src/pages/ExperimentQuick.jsx:21]: >  Currently if i select job A on page 1 and go to page 2 the job is deselected. Is it intended?

Yes, it is intended. We could discuss if this is the behavior we want.

>  I attach a video, not sure if it's worth fixing:

I see. Let me check it because it should not be happening for sure.

File: src/pages/ExperimentQuick.jsx (hunk 10 above)
Comment (pablogoitia) [reply src/pages/ExperimentQuick.jsx]: In this specific case it is a bug. The selector is meant to be hidden when the items are less than 100. This is failing because of a last hour change I did, so it will be easy to revert

File: cypress/e2e/quick_view.cy.js (hunk 12 above)
Comment (pablogoitia) [reply cypress/e2e/quick_view.cy.js:311]: Yes! I understand the surprise :) In this case I just needed `page_items` to be one because it is the parameter that the page size selector uses (the part I am testing). Whether the list appears empty does not affect this test. The idea behind was just to save some lines

File: src/pages/ExperimentQuick.jsx (hunk 11 above)
Comment (pablogoitia) [reply src/pages/ExperimentQuick.jsx:281]: Thanks @ntorqulu! I added the proposal here: https://github.com/BSC-ES/autosubmit-gui/issues/460#issuecomment-5696681955

File: cypress/support/utils.js (hunk 13 above)
Comment (pablogoitia) [reply cypress/support/utils.js:1]: Honestly I had doubts when I did this. Initially I imported it directly from utils, but it's true that if a constant in the utils is mistakenly changed, the test would pass anyway. I don't mind using one or another, because the probability of this to happen is very low, but I think having this double check is effortless

File: src/pages/ExperimentQuick.jsx (hunk 9 above)
Comment (pablogoitia) [reply src/pages/ExperimentQuick.jsx:21]: Hi, @ntorqulu. I have found the problem. I do not know why but the `selectedJobIds` was being defined in both components, producing this misbehavior. I have left it in the "parent" component and passed it to the "child".

I will upload the changes soon.

File: public/CHANGELOG.md [hunk 22]
```diff
@@ -2,6 +2,7 @@
 
 ### Pre-release v4.1.1-beta.5 - Unreleased
 * **Compatibility with Autosubmit API version >= [v4.1.2b7](https://pypi.org/project/autosubmit-api/v4.1.2b7/)** (unreleased)
+* Now the Quick View supports pagination and lazy loading.
```
Comment (ntorqulu) [inline public/CHANGELOG.md:5]: It only supports pagination right? Is a pagination on demand of the job list.

File: src/pages/ExperimentQuick.jsx [hunk 23]
```diff
@@ -168,92 +220,104 @@ const ExperimentQuick = () => {
...
+          </button>
         </div>
-        <div className="text-sm" style={{ whiteSpace: "nowrap" }}>
-          Showing {jobs.length} of{" "}
-          <strong>
-            {data && data.tree_view && data.tree_view.length} total jobs
-          </strong>
+        <div className="flex items-center gap-1 text-sm" style={{ whiteSpace: "nowrap" }}>
+          <span>Showing</span>
+          {data?.pagination?.total_items <= Math.min(...ITEMS_PER_PAGE_OPTIONS) ? (
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx:302]: `Math.min(...)` is computed every render. Can we assume the array is sorted? `ITEMS_PER_PAGE_OPTION` will always have as first element the `DEFAULT_ITEMS_QUICK_VIEW` no?
Can we optimize it a little bit and do `ITEMS_PER_PAGE_OPTIONS[0]` instead?

File: src/pages/ExperimentQuick.jsx [hunk 24]
```diff
@@ -1,15 +1,25 @@
...
 import FetchJobDetailCard from "../common/FetchJobDetailCard";
+import Paginator from "../common/Paginator";
+
+const ITEMS_PER_PAGE_OPTIONS = [DEFAULT_ITEMS_QUICK_VIEW, 500, 1000];
+const DEFAULT_STATUS_QUICK_VIEW = "Any status";
+
+const QuickJobList = ({ jobs, selectedJobIds, onSelectionChange }) => {
+  const [lastClickedIndex, setLastClickedIndex] = useState(null);
+
+  useEffect(() => {
```
Comment (ntorqulu) [inline src/pages/ExperimentQuick.jsx:19]: Now that the ownership has moved to the parent, this effect can be removed? QuickJobList doesn't need to write to the parent now.

**Review by ntorqulu — APPROVED** [review]
Hi @pablogoitia, last review!
Just comments about cleaning dead code, I leave it approved.
I've been testing it with the API and I've found no user problems from my side, the multi-selection across pages is working correctly now!

**Comment by pablogoitia** [comment]
Hi, @ntorqulu. As you see, I needed to push some corrections alongside your suggestions, so here I summarize the changes:

* Const `DEFAULT_ITEMS_PER_PAGE` now belongs to `ExperimentQuick.jsx`, not `utils`.
* Now `QuickJobList` is not the responsible for resetting the status of either `selectedJobIds` or `lastClickedIndex`.
* `handleJobClick` -> refactor
* The selected job ids and the last selected job index are now grouped within a dictionary called `selection`, in the parent component. This is because the job ids are needed in both components, and the index must be reset at the same time as the selected job list. This is a intermediate solution between what we had before and creating a dedicated hook to be shared between the parent and child components (which does not make sense for me because is a very specific functionality of the quick view, at least for now).
* Added _docstrings_.

Let me know if I still have the approval :)

**Comment by ntorqulu** [comment]
Mmm @pablogoitia, only did a quick test but I don't think this should happen:

[Screencast from 2026-09-22 11-52-46.webm](https://github.com/user-attachments/assets/4be457b6-0747-4c41-b1f0-0df88716aef0)

If I select three jobs on the first page, change page, and select only one in the second page the position of the first three are also selected.
I think that in page 2 only the jobs I click should be selected...

**Comment by pablogoitia** [comment]
Hi, @ntorqulu. Fortunately this is the expected behavior. It is not a matter of a erroneous shared state between pages, but how the index reset works:

Before, the last clicked job index was always initialized to null. This meant that if you firstly select a job with the shift key, it would just select that job. However, now the index is initialized to zero. This means that by firstly clicking a job with shift, all the jobs from the beginning of the list to that job will be selected.

This is a change introduced in this last commit because I found it to be a more natural behavior. For example, try to do the same within your Linux file explorer, or even in the tree view of the GUI.

I think this is a matter of taste, because opinions on what is expected to be natural for users here may be very diverse.

Let me know what do you think

**Comment by ntorqulu** [comment]
Hahah yes, I realized that it was not as unexpected as I thought. It's just that I was used to another behavior (which is not the usual one). Thanks!

**Review by ntorqulu — APPROVED** [review]
Hi @pablogoitia, all good from my side. Thanks for changing the job selection behavior, now it's more predictable!
Also, +1 to adding doctrings, maybe we could also configure ESLint and prettier to ensure consistency and catch bugs earlier? I can open an issue for this.

**Comment by pablogoitia** [comment]
> Also, +1 to adding doctrings, maybe we could also configure ESLint and prettier to ensure consistency and catch bugs earlier? I can open an issue for this.

Feel free to do it. I found somewhat hard to expand this implementation because there were no comments 😅️

Thanks a lot for taking that much time for the reviews, they were very helpful 🙇🏻‍♂️️
